components/sup/src/manager/mod.rs

```diff
@@ -720,18 +728,20 @@ impl Manager {
                 self.take_action_on_services()?;
             }
 
-            self.update_peers_from_watch_file()?;
             self.update_running_services_from_user_config_watcher();
             self.check_for_updated_packages();
-            self.restart_elections();
-            self.census_ring.update_from_rumors(
-                &self.butterfly.service_store,
-                &self.butterfly.election_store,
-                &self.butterfly.update_store,
-                &self.butterfly.member_list,
-                &self.butterfly.service_config_store,
-                &self.butterfly.service_file_store,
-            );
+            if let Some(ref butterfly) = self.butterfly {
+                Self::update_peers_from_watch_file(butterfly, &self.peer_watcher)?;
```

Comment:
I haven't dug in enough to fully grok the reasons why, but off the cuff, it strikes me as odd that this went from a struct level function/method to a type level function.  What was the reasoning behind that?

Reply:
Ah, good catch. There was a borrow error because I wasn't doing the pattern matching properly on line 733. Changing the function was the only way I could figure out to make it work. I fixed the bit on 733 though, so this shouldn't be necessary.

Reply:
Now that we're changing `self.butterfly` to be an `Option`, I think this can be a good pattern to avoid having `if let Some(…) = self.butterfly` at all levels of the code and passing only the specific data functions *need* into associated functions (i.e., `Self::f()`) rather than using methods  (i.e., `self.f()`) which carry around superfluous state and require more places where we encounter the `Option`. See https://github.com/habitat-sh/habitat/pull/6176#discussion_r258524765.